### ISSUE TITLE
Improve attribute hiding pattern

### ISSUE THREAD

**Issue by xaviernogueira** [xaviernogueira/xpublish-file-metadata#1]
As @abkfenris suggested in 10/6/2023 Xpublish meeting, the current attribute hiding pattern leaves more to be desired, especially for users migrating from ERDAP.

@abkfenris please add comments to this issue with your perspective once you get the chance! Also tagging @jhamman, @mpiannucci, @jonmjoyce 

From what I remember:
* We want dataset specific attribute hiding, simular to how we have it based on file format but using `dataset_id`.
* Better naming for global attribute hiding vs file format attribute hiding vs dataset specific attribute hiding

Since the renaming may break backward compatibility, this should be done before the first version release.

**Comment by abkfenris** [comment]
ERDDAP can both add and remove attributes from datasets as well as do some creative deriving. Probably don't need all of that functionality, but here are [ERDDAP's docs for `<addAttribute>`](https://coastwatch.pfeg.noaa.gov/erddap/download/setupDatasetsXml.html#addAttributes).

I would probably break things out so that global/file type/per dataset had it's own plugin field, so maybe altering the plugin signature to something like:

```py
class FileMetadataPlugin(Plugin):
    """File metadata plugin for xpublish."""

    name: str = 'file-metadata'

    dataset_router_prefix: str = '/file-metadata'
    dataset_router_tags: Sequence[str] = ['file-metadata']

    hide_attrs_global: list[str] = Field(default_factory=list)
    hide_attrs_by_file_type: dict[FileFormats, listr[str]] = Field(default_factory=dict)
    hide_attrs_by_dataset_pattern: dict[str, list[str]] = Field(default_factory=dict)
```

Then it could be used like:

```py
FileMetadataPlugin(
    hide_attrs_global=["this_key_tells_another_plugin_to_do_something_cool"],
    hide_attrs_by_file_type={"grib": ["we_are_really_ashamed_of_using_this_tool_on_gribs"]},
    hide_attrs_by_dataset_pattern={"*01_met": ["the_met_logger_on_this_group_of_buoys_are_weird"]}
)
```

I'm thinking that it might be nice to enable matching datasets via glob (well [`fnmatch.fnmatch`](https://docs.python.org/3/library/fnmatch.html#fnmatch.fnmatch)) or regex patterns, as there can often be a set of datasets from similar processing pipelines that might have the same artifacts to be hidden.

Maybe the dataset specific hiding could wait for a later release. However, it would be worth pondering at this point, do the attrs to be hidden get merged, or does the most specific match's set of attrs to hid 'win'? Right now it looks like the attrs to be hidden are pre-computed for each format.

One question is how does this work on attributes on a variable? Can they be specified to be hidden?

It may also be worth looking at what Intake does. Axiom has a [repo](https://github.com/axiom-data-science/mc-goods/) that has a lot of examples of tweaking metadata, but I'm not sure if they or Intake in general can remove attributes, and I'm having a hard time finding that in the Intake docs at the moment.

**Comment by xaviernogueira** [comment]
@abkfenris Great thank you for the details! I can definitely use those different plugin attributes, and a pattern match is a great idea. I will add both features.

Regarding variable attributes those are not explicitly found, as it is reading the NetCDF/GeoTIFF file metadata at a top level via the identifying the path via the encoding attributes on xarray, and then attempting to read with NetCDR4 or rasterio. I can definitely explore this idea, and moved it to issue #4 .

Another change before it is properly usable is having a different pattern (if necessary) for reading metadata on bucket hosted data, but I will leave that in another issue #3 .
